content/blog/2024-01-11-xilem-backend-roadmap.md

```diff
@@ -0,0 +1,157 @@
+
+If we do move on in that direction, then Masonry will need some porting work: 
+
+- Using Vello and Parley instead of Piet.
+- Using Glazier instead of druid-shell.
+- Integrating AccessKit.
+- Removing the dependency on GTK.
+
+AccessKit support in particular is table stakes, by now. The Rust ecosystem is converging on the idea that accessibility (or at least, screen reader support) should be integrated early on in the design of GUI frameworks, not tacked on at the end.
+
+There are other things Xilem added over Druid (notably virtual lists and async support), but I believe Masonry doesn't need to implement them right away for parity. Both will be very important for eventual performance work.
+
+
+### Not using custom Widgets
+
+This is another big one, and will likely also require an RFC.
+
+We should write Xilem's backend under the assumption that end users of the library (including the Xilem frontend) won't be able to create their own widgets. Instead, they will have to compose the primitives given to them the same way they compose DOM elements in the browser. The primary escape hatch will be a fixed-size surface where end-users can use arbitrary paint primitives, like the web canvas.
+
+This is a bold move, but one I'm confident in. Developers of hobby GUI frameworks tend to provide a severely anemic set of primitives, under the assumption that if anyone wants other features in their widgets, they can implement them themselves. Both Druid and Xilem suffer from this.
+
+On the other end of the spectrum, the web has shown that you can implement complex custom UIs given a rigid but rich set of primitives.
+
+One major customization point I want to remove from Xilem is layout.
+
+Today all Rust GUI frameworks ship their own layout algorithm, but none of them are *competing* on their layout algorithm. Often the layout algorithm is just a means to an end; developers that do want to work on a state-of-the-art layout engine will provide it in a modular crate like [Morphorm](https://github.com/vizia/morphorm) and [Taffy](https://github.com/DioxusLabs/taffy).
+
+And here's the thing: realistically, most layout engines aren't that creative. Every Rust framework and every major GUI framework I'm aware of uses layout that boils down to "FlexBox with small variations". The number of meaningfully different ways to do GUI layout can be counted on, generously, two hands.
+
+I'd argue we only need CSS Flow, FlexBox, CSS Grid, and maybe Morphorm to cover most people's needs. Conveniently, this is what Taffy aims to provide. There's already an experimental Taffy integration in Xilem, and I want to make it deeper. More on this in a latter blog post.
```

Comment:
I think it's pretty cool, that we can provide taffy layout without forcing it into the core. But it remains to be seen, whether keeping the exact layout computation open to the widget, it could well be that this unnecessarily complexifies things (e.g. because of multiple passes) and this may be handled in taffy instead. I do sometimes think (though admittedly rarely) that current layout engines are limiting. (Interesting paper btw.: [ORC Layout](https://duruofei.com/papers/Jiang_ORCLayout-AdaptiveGUILayoutWithOR-Constraints_CHI2019.pdf))

Reply:
I may need to rewrite this to be more explicit; but my goal here is to make Taffy a lot closer to the core than it currently is in Xilem.

Reply:
I'll read the ORC paper, thanks!

Reply:
Yes I undestood that, and I think it's already clear. I'm just not yet convinced that this is necessarily a good idea, as it makes fancier layout more difficult or locks it more into web-like layouting, but I could of course be wrong. I think we had a longer discussion about this topic (shortly after the taffy layout PR popped up) some while back in office hours, I'm not sure if you were there.